apps/server/test/lexical/server/code_intelligence/completion/translations/macro_test.exs

```diff
@@ -46,6 +46,63 @@ defmodule Lexical.Server.CodeIntelligence.Completion.Translations.MacroTest do
       assert apply_completion(completion) == "def ${1:name}($2) do\n  $0\nend"
     end
 
+    test "def preceeded by a @spec with args", %{project: project} do
+      source = ~q[
+        @spec my_function(term(), term()) :: term()
+        def|
+      ]
+
+      assert {:ok, completion} =
+               project
+               |> complete(source)
+               |> fetch_completion("def ")
+
+      assert apply_completion(completion) == ~q[
+        @spec my_function(term(), term()) :: term()
+        def my_function(${1:arg_1}, ${2:arg_2}) do
+          $0
+        end
+      ]
+    end
+
+    test "def preceeded by a @spec without args", %{project: project} do
+      source = ~q[
+        @spec my_function :: term()
+        def|
+      ]
+
+      assert {:ok, completion} =
+               project
+               |> complete(source)
+               |> fetch_completion("def ")
+
+      assert apply_completion(completion) == ~q[
+        @spec my_function :: term()
+        def my_function do
+          $0
+        end
+      ]
+    end
+
+    test "defp preceeded by a @spec with args", %{project: project} do
+      source = ~q[
+        @spec my_function(term(), term()) :: term()
+        def|
+      ]
+
+      assert {:ok, completion} =
+               project
+               |> complete(source)
+               |> fetch_completion("defp ")
+
+      assert apply_completion(completion) == ~q[
+        @spec my_function(term(), term()) :: term()
+        defp my_function(${1:arg_1}, ${2:arg_2}) do
+          $0
+        end
+      ]
+    end
+
```

Comment:
how about a spec like `@spec my_function(first :: term(), second :: term()) :: term()

Reply:
I'm not completely sure which case this suggestion applies to. Do you mean that we could use existing function parameter names when generating specs, e.g.

```elixir
@spe|
def my_function(x, y, z) do
  ...
end

# becomes

@spec my_function(x :: term(), y :: term(), z :: term()) :: term()
def my_function(x, y, z) do
  ...
end
```

I think this is worth doing, though we won't always be able to infer the name (we can handle the cases `var`, `var = PATTERN`, `PATTERN = var`).

Or did you mean something else?

Reply:
The opposite is also possible:

```elixir
@spec my_function(x :: term(), y :: term(), z :: term()) :: term()
def|

# becomes

@spec my_function(x :: term(), y :: term(), z :: term()) :: term()
def my_function(x, y, z) do
  |
end
```

Reply:
You understood my meaning. Inferring the names of the arguments seems like a win.